drivers/azure/azureutil/securityrules.go

```diff
@@ -0,0 +1,319 @@
+			Description:              to.StringPtr("Allow docker engine access (TLS-protected)"),
+			SourceAddressPrefix:      to.StringPtr("*"),
+			DestinationAddressPrefix: to.StringPtr("*"),
+			SourcePortRange:          to.StringPtr("*"),
+			DestinationPortRange:     to.StringPtr(fmt.Sprintf("%d", dockerPort)),
+			Access:                   network.SecurityRuleAccessAllow,
+			Direction:                network.SecurityRuleDirectionInbound,
+			Protocol:                 network.SecurityRuleProtocolTCP,
+			Priority:                 to.Int32Ptr(dockerPriority),
+		},
+	})
+	return &newRules, nil
+}
+
+func syncOpenPorts(rules *[]network.SecurityRule, openPorts []string) (*[]network.SecurityRule, error) {
+	openPortsToBeDeleted := make(map[string]network.SecurityRule)
+	takenPriorities := make(map[int32]bool)
+	// Add all rules for deletion by default
+	for _, rule := range *rules {
+		takenPriorities[to.Int32(rule.Priority)] = true
+		if isOpenPort(rule) {
+			openPortsToBeDeleted[to.String(rule.Name)] = rule
+		}
+	}
+	// Remove from deletion if it exists or add to queue if it does not exist
+	openPortsToBeAdded := make(map[string]network.SecurityRule)
+	for _, p := range openPorts {
+		openPort, err := parseOpenPort(p)
+		if err != nil {
+			return nil, err
+		}
+		if _, ok := openPortsToBeDeleted[openPort.Name]; ok {
+			// Do not delete the open port
+			delete(openPortsToBeDeleted, openPort.Name)
+			continue
+		}
+		openPortsToBeAdded[openPort.Name] = network.SecurityRule{
+			Name: to.StringPtr(openPort.Name),
+			SecurityRulePropertiesFormat: &network.SecurityRulePropertiesFormat{
+				Description:              to.StringPtr("User requested port to be accessible from Internet via docker-machine"),
+				SourceAddressPrefix:      to.StringPtr("*"),
+				DestinationAddressPrefix: to.StringPtr("*"),
+				SourcePortRange:          to.StringPtr("*"),
+				DestinationPortRange:     to.StringPtr(openPort.Port),
+				Access:                   network.SecurityRuleAccessAllow,
+				Direction:                network.SecurityRuleDirectionInbound,
+				Protocol:                 openPort.Proto,
+			},
+		}
+	}
+	// Log open ports to be deleted and added
+	if len(openPortsToBeDeleted) > 0 || len(openPortsToBeAdded) > 0 {
+		removedPorts := make([]string, len(openPortsToBeDeleted))
+		addedPorts := make([]string, len(openPortsToBeAdded))
+		i := 0
+		for openPort := range openPortsToBeDeleted {
+			removedPorts[i] = openPort
+		}
+		i = 0
+		for openPort := range openPortsToBeAdded {
+			addedPorts[i] = openPort
+		}
+		log.Info("Updating security rule configuration", logutil.Fields{
+			"openPortsRemoved": removedPorts,
+			"openPortsAdded":   addedPorts,
+		})
+	}
+	// Remove all unnecessary open ports and release the priorities
+	for _, rule := range openPortsToBeDeleted {
+		delete(takenPriorities, to.Int32(rule.Priority))
+		rules = removeSecurityRule(rules, rule)
+	}
+	// Add any necessary open ports
+	newRules := *rules
```

Comment:
similar conversation as addSshPort and addDockerPort?

Reply:
updated